packages/marketplace/contracts/TransferManager.sol

```diff
@@ -199,18 +206,101 @@ abstract contract TransferManager is Initializable, ITransferManager {
         DealSide memory paymentSide,
         DealSide memory nftSide
     ) internal returns (uint256) {
-        (address token, uint256 tokenId) = LibAsset.decodeToken(nftSide.asset.assetType);
-        IRoyaltiesProvider.Part[] memory royalties = royaltiesRegistry.getRoyalties(token, tokenId);
+        if (nftSide.asset.assetType.assetClass == LibAsset.AssetClass.BUNDLE) {
+            LibAsset.Bundle memory bundle = LibAsset.decodeBundle(nftSide.asset.assetType);
+
+            for (uint256 i; i < bundle.bundledERC721.length; i++) {
+                address token = bundle.bundledERC721[i].erc721Address;
+                uint256 idLength = bundle.bundledERC721[i].ids.length;
+                for (uint256 j; j < idLength; j++) {
+                    IRoyaltiesProvider.Part[] memory royalties = royaltiesRegistry.getRoyalties(
+                        token,
+                        bundle.bundledERC721[i].ids[j]
+                    );
+
+                    remainder = _applyRoyalties(
+                        remainder,
+                        paymentSide,
+                        bundle.priceDistribution.erc721Prices[i][j],
+                        royalties,
+                        nftSide.recipient
+                    );
+                }
+            }
+
+            for (uint256 i; i < bundle.bundledERC1155.length; i++) {
+                address token = bundle.bundledERC1155[i].erc1155Address;
+                uint256 idLength = bundle.bundledERC1155[i].ids.length;
+                require(idLength == bundle.bundledERC1155[i].supplies.length, "ERC1155 array error");
+                for (uint256 j; j < idLength; j++) {
+                    IRoyaltiesProvider.Part[] memory royalties = royaltiesRegistry.getRoyalties(
+                        token,
+                        bundle.bundledERC1155[i].ids[j]
+                    );
+                    remainder = _applyRoyalties(
+                        remainder,
+                        paymentSide,
+                        bundle.priceDistribution.erc1155Prices[i][j],
+                        royalties,
+                        nftSide.recipient
+                    );
+                }
+            }
+
+            uint256 quadSize = bundle.quads.xs.length;
+            if (quadSize > 0) {
+                for (uint256 i = 0; i < quadSize; i++) {
+                    uint256 size = bundle.quads.sizes[i];
+                    uint256 x = bundle.quads.xs[i];
+                    uint256 y = bundle.quads.ys[i];
+
+                    uint256 tokenId = idInPath(0, size, x, y);
```

Comment:
Can you explain why do we use `0` here?

Reply:
So for calculating royalty of a quad, we fetch royalty of top-left land ( at index 0).
That is then used to calculate royalty on entire quad price

Reply:
Got it, thats what I thought, thanks!